pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -337,32 +340,44 @@ func (configgen *ConfigGeneratorImpl) buildGatewayInboundHTTPRouteConfig(
 
 	port := int(servers[0].Port.Number)
 	// NOTE: WE DO NOT SUPPORT two gateways on same workload binding to same virtual service
-	virtualServices := env.VirtualServices(gateways)
+	allVirtualServices := env.VirtualServices(gateways)
+	virtualServices := mergeIngress(allVirtualServices)
+
 	virtualHosts := make([]route.VirtualHost, 0, len(virtualServices))
+	dedupHosts := map[string]*model.Config{}
 	for _, v := range virtualServices {
 		vs := v.Spec.(*networking.VirtualService)
 		matchingHosts := pickMatchingGatewayHosts(gatewayHosts, vs.Hosts)
 		if len(matchingHosts) == 0 {
-			log.Warnf("omitting virtual service %q because its hosts don't match gateways %v server %d", v.Name, gateways, port)
+			log.Infof("%s omitting virtual service %q because its hosts  don't match gateways %v server %d", node.ID, v.Name, gateways, port)
 			continue
 		}
-		routes, err := istio_route.BuildHTTPRoutesForVirtualService(v, svcs, port, nil, gateways, env.IstioConfigStore)
+		routes, err := istio_route.BuildHTTPRoutesForVirtualService(*v, svcs, port, nil, gateways, env.IstioConfigStore)
 		if err != nil {
-			log.Warnf("omitting routes for service %v due to error: %v", v, err)
+			log.Warnf("%s omitting routes for service %v due to error: %v", node.ID, v, err)
 			continue
 		}
 
 		for vsvcHost, gatewayHost := range matchingHosts {
-			host := route.VirtualHost{
-				Name:    fmt.Sprintf("%s:%d", v.Name, port),
-				Domains: []string{vsvcHost},
-				Routes:  routes,
-			}
-
-			if tlsRedirect[gatewayHost] {
-				host.RequireTls = route.VirtualHost_ALL
+			key := fmt.Sprintf("%s:%d", v.Name, port)
+			old, f := dedupHosts[key]
+			if f {
+				push.Add(model.DuplicatedDomains, key, node,
+					fmt.Sprintf("Duplicated virtual host %s use %s:%s rejecting %s:%s",
+						key, old.Namespace, old.Name, v.Namespace, v.Name))
```

Comment:
I don't think this is that simple (string match). Gateway's allow wildcards and we respect wildcards and absorb hosts under a given wildcard. The issues you were seeing were mostly due to the incorrect merging you attempt below. As such, you shouldn't be hitting any case of duplicate virtual hosts unless you generate two virtual services for saem ingress host, in conversion.go. So I would check the logic there.